From 0b38bfce7dc9be0d4273e0079037c7326e654bf0 Mon Sep 17 00:00:00 2001
From: Kyle Hsiao <kylehsiao@google.com>
Date: Tue, 26 May 2026 20:06:56 -0700
Subject: [PATCH 03/13] Fix potential integer overflow in Type 3 Tag block
 operations.

The block_id variable is changed to uint32_t to handle block IDs up to 65535. The bounds checks and loop conditions in rw_t3t_send_next_ndef_update_cmd and rw_t3t_send_next_ndef_check_cmd are updated to perform calculations using uint32_t to prevent integer overflow when determining the range of blocks to write or check. The command buffer is now freed on error in the bounds check.

Bug: 503544522
Test: Manual
Flag: EXEMPT BUGFIX
Change-Id: I53765774042218c3fd25985e7da64ccbe9bc57fd
Merged-In: I53765774042218c3fd25985e7da64ccbe9bc57fd
(cherry picked from commit 241c49442390509f106f711ffcf11a476508898a)
---
 libnfc-nci/src/nfc/tags/rw_t3t.cc | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/libnfc-nci/src/nfc/tags/rw_t3t.cc b/libnfc-nci/src/nfc/tags/rw_t3t.cc
index 5a16703..ad60936 100644
--- a/libnfc-nci/src/nfc/tags/rw_t3t.cc
+++ b/libnfc-nci/src/nfc/tags/rw_t3t.cc
@@ -771,7 +771,6 @@ tNFC_STATUS rw_t3t_send_update_ndef_attribute_cmd(tRW_T3T_CB* p_cb,
 *****************************************************************************/
 tNFC_STATUS rw_t3t_send_next_ndef_update_cmd(tRW_T3T_CB* p_cb) {
   tNFC_STATUS retval = NFC_STATUS_OK;
-  uint16_t block_id;
   uint16_t first_block_to_write;
   uint16_t ndef_blocks_to_write, ndef_blocks_remaining;
   uint32_t ndef_bytes_remaining, ndef_padding = 0;
@@ -780,7 +779,7 @@ tNFC_STATUS rw_t3t_send_next_ndef_update_cmd(tRW_T3T_CB* p_cb) {
   NFC_HDR* p_cmd_buf;
   uint8_t *p_cmd_start, *p;
   uint8_t blocks_per_update;
-  uint32_t timeout;
+  uint32_t timeout, block_id;
 
   p_cmd_buf = rw_t3t_get_cmd_buf();
   if (p_cmd_buf != nullptr) {
@@ -847,7 +846,8 @@ tNFC_STATUS rw_t3t_send_next_ndef_update_cmd(tRW_T3T_CB* p_cb) {
         ndef_blocks_to_write); /* Number of blocks to write in this command */
     timeout = rw_t3t_update_timeout(ndef_blocks_to_write);
 
-    if (ndef_blocks_to_write > UINT16_MAX - first_block_to_write) {
+    if ((uint32_t)first_block_to_write + ndef_blocks_to_write > 0x10000) {
+        GKI_freebuf(p_cmd_buf);
         LOG(ERROR) << StringPrintf("%s: Heap out-of-bounds",
                                __func__);
         android_errorWriteLog(0x534e4554, "508389055");
@@ -855,7 +855,7 @@ tNFC_STATUS rw_t3t_send_next_ndef_update_cmd(tRW_T3T_CB* p_cb) {
     }
 
     for (block_id = first_block_to_write;
-         block_id < (first_block_to_write + ndef_blocks_to_write); block_id++) {
+         block_id < ((uint32_t)first_block_to_write + ndef_blocks_to_write); block_id++) {
       if (block_id < 256) {
         /* Block IDs 0-255 can be specified in '2-byte' format: byte0=0,
          * byte1=blocknumber */
@@ -929,9 +929,8 @@ tNFC_STATUS rw_t3t_send_next_ndef_update_cmd(tRW_T3T_CB* p_cb) {
 *****************************************************************************/
 tNFC_STATUS rw_t3t_send_next_ndef_check_cmd(tRW_T3T_CB* p_cb) {
   tNFC_STATUS retval = NFC_STATUS_OK;
-  uint16_t block_id;
   uint16_t ndef_blocks_remaining, first_block_to_read, cur_blocks_to_read;
-  uint32_t ndef_bytes_remaining;
+  uint32_t ndef_bytes_remaining, block_id;
   NFC_HDR* p_cmd_buf;
   uint8_t *p_cmd_start, *p;
 
@@ -996,7 +995,8 @@ tNFC_STATUS rw_t3t_send_next_ndef_check_cmd(tRW_T3T_CB* p_cb) {
     UINT8_TO_STREAM(
         p, cur_blocks_to_read); /* Number of blocks to check in this command */
 
-    if (cur_blocks_to_read > UINT16_MAX - first_block_to_read) {
+    if ((uint32_t)first_block_to_read + cur_blocks_to_read > 0x10000) {
+        GKI_freebuf(p_cmd_buf);
         LOG(ERROR) << StringPrintf("%s: Heap out-of-bounds",
                                __func__);
         android_errorWriteLog(0x534e4554, "508389055");
@@ -1004,7 +1004,7 @@ tNFC_STATUS rw_t3t_send_next_ndef_check_cmd(tRW_T3T_CB* p_cb) {
     }
 
     for (block_id = first_block_to_read;
-         block_id < (first_block_to_read + cur_blocks_to_read); block_id++) {
+         block_id < ((uint32_t)first_block_to_read + cur_blocks_to_read); block_id++) {
       if (block_id < 256) {
         /* Block IDs 0-255 can be specified in '2-byte' format: byte0=0,
          * byte1=blocknumber */
-- 
2.55.0

